R/ggridges.R

```diff
@@ -0,0 +1,289 @@
+    prestats_data, layout, params, p, 
+    closed = FALSE, 
+    ...
+  )
+  
+}
+
+
+#' @export
+to_basic.GeomDensityRidges2 <- function(data, prestats_data, layout, params, p, ...) {
+  
+  to_basic(
+    prefix_class(data, 'GeomDensityRidgesGradient'),
+    prestats_data, layout, params, p, 
+    closed = TRUE, 
+    ...
+  )
+  
+}
+
+
+
+#' @export
+to_basic.GeomDensityLine <- function(data, prestats_data, layout, params, p, ...) {
+  
+  to_basic(prefix_class(data, 'GeomDensity'))
+  
+}
+
+
+
+#' @export
+to_basic.GeomRidgeline <- function(data, prestats_data, layout, params, p, ...) {
+
+  to_basic(
+    prefix_class(data, 'GeomDensityRidgesGradient'),
+    prestats_data, layout, params, p, ...
+  )
+
+}
+
+
+# TODO: Implement following
+to_basic.GeomRidgelineGradient <- function(data, prestats_data, layout, params, p, ...) {
+
+
+  d <- get_ridge_data(data, params$na.rm)
+
+  # split data into separate groups
+  groups <- split(d, factor(d$group))
+
+  # sort list so lowest ymin values are in the front (opposite of ggridges)
+  o <- order(
+    unlist(
+      lapply(
+        groups,
+        function(data) data$ymin[1]
+      )
+    ),
+    decreasing = FALSE
+  )
+  groups <- groups[o]
+
+  # for each group create a density + vline + point as applicable
+  res <- lapply(
+    rev(groups),
+    function(x){
+
+      draw_stuff <- split(x, x$datatype)
+
+      # first draw the basic density ridge part
+
+      stopifnot(!is.null(draw_stuff$ridgeline))
+      d2 <- d1 <- draw_stuff$ridgeline
+      d2$colour <- NA # no colour for density area
+      d2$fill_plotlyDomain <- NA
+
+      d1$y <- d1$ymax
+      d1$alpha <- 1 # don't use fill alpha for line alpha
+
+      # calculate all the positions where the fill type changes
+      fillchange <- c(FALSE, d2$fill[2:nrow(d2)] != d2$fill[1:nrow(d2)-1])
+
+      # and where the id changes
+      idchange <- c(TRUE, d2$group[2:nrow(d2)] != d2$group[1:nrow(d2)-1])
+
+      # make new ids from all changes in fill style or original id
+      d2$ids <- cumsum(fillchange | idchange)
+
+      # get fill color for all ids
+      fill <- d2$fill[fillchange | idchange]
+
+      # rows to be duplicated
+      dupl_rows <- which(fillchange & !idchange)
+      d2$y <- d2$ymax
+      if (length(dupl_rows)>0){
+        rows <- d2[dupl_rows, ]
+        rows$ids <- d2$ids[dupl_rows-1]
+        rows <- rows[rev(seq_len(nrow(rows))), , drop = FALSE]
+        # combine original and duplicated d2
+        d2 <- rbind(d2, rows)
+      }
+
+      # split by group to make polygons
+      # browser()
+      d2 <- tibble::deframe(tidyr::nest(d2, .by = 'ids'))
+
+      ridges <- c(
+        d2,
+        list(
+          to_basic(prefix_class(d1, "GeomLine"))
+        )
+      )
+
+      ridges
+    }
+  )
+  # set list depth to 1
+  unlist(res, recursive = FALSE)
+
+
+}
+
+
+
+#' @export
+geom2trace.GeomRidgelineGradient <- function(data, params, p) {
+  
+  # data <- group2NA(data)
```

Comment:
Are uncertain whether or not this is needed? 

(same question for the commented out lines below)

Reply:
removed commented code